src/dcap_provider.cpp

```diff
@@ -236,24 +247,75 @@ static inline quote3_error_t fill_qpl_string_buffer(
 }
 
 //
-// determines the maximum age in local cache
+// Determine time cache should invalidate for given collateral
 //
-sgx_plat_error_t get_cache_max_age(const curl_easy&, time_t* max_age)
+bool get_cache_expiration_time(CollateralTypes collateral_type, time_t &expiration_time)
 {
-    if (max_age == nullptr)
+    time_t max_age = 0;
+    tm* max_age_s = localtime(&max_age);
+
+    switch(collateral_type)
     {
-        return SGX_PLAT_ERROR_INVALID_PARAMETER;
+        case CollateralTypes::TcbInfo:
+        case CollateralTypes::QeIdentity:
+        case CollateralTypes::QveIdentity:
+        {
+            max_age_s->tm_hour += 12;

```

Comment:
There is a potential issue - local cache may not be aware an collateral has expired. For example, at 12:00 PM we locally cache a QE Identity what will expire at 1:00 PM on the same day and this QE Identity will stay in local cache for 12 hours. From 1:00 PM to midnight, the attestation will fail due to QE Identity expiration.

Reply:
Since PCK cache will typically update the collateral well before expiration, I think we will be OK here but you are absolutely right I'll fix this

Reply:
I am not sure how to fix the issue unless we parse the collaterals?

---
In reply to: [423988419](https://github.com/microsoft/Azure-DCAP-Client/pull/105#discussion_r423988419) [](ancestors = 423988419)

Reply:
Yes you are correct - in the interest of performance and simplicity I think it may be better to rely on the PCK cache update policy. I just confirmed the settings for QE ID and CRLS - they get refreshed 5 days before expiration